Unsaved questions param values

dpsutton opened the pull request:
Allow for two endpoints to service parameter values for use during dashboard and card creation. The key distinguisher here is that these cards or dashboards do not yet exist. There is no check to do against the the parameters saved on the card.

- `POST api/dataset/parameter/values`
- `POST api/dataset/parameter/search/foo`

> :warning:
> The api response does not currently deduplicate values from these queries. We will solve that in a follow up PR to do it comprehensively and in the correct location.

#### STATIC LIST VALUES

```
❯ http post "localhost:3000/api/dataset/parameter/values" \
   parameter:='{"values_source_type": "static-list", "values_source_config": {"values": ["foo", "bar"]}}' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        "foo",
        "bar"
    ]
}
```

#### STATIC LIST SEARCH

```
❯ http post "localhost:3000/api/dataset/parameter/search/fo" \
   parameter:='{"values_source_type": "static-list", "values_source_config": {"values": ["foo", "bar"]}}' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        "foo"
    ]
}
```

#### CARD VALUES

```
❯ http post "localhost:3000/api/dataset/parameter/values" \
   parameter:='{"values_source_type": "card", "values_source_config": {"card_id": 4, "value_field": ["field",58,null]}}' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        "Gizmo",
        "Doohickey",
        "Doohickey",
        ...
        "Gizmo",
        "Widget",
        "Gadget"
    ]
}
```

#### CARD SEARCH

```
❯ http post "localhost:3000/api/dataset/parameter/search/ga" \
   parameter:='{"values_source_type": "card", "values_source_config": {"card_id": 4, "value_field": ["field",58,null]}}' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        "Gadget",
        "Gadget",
        "Gadget",
        "Gadget",
        "Gadget",
        ...
        "Gadget",
        "Gadget",
        "Gadget"
    ]
}
```

#### FIELD VALUES

```
❯ http post "localhost:3000/api/dataset/parameter/values" \
   parameter:='{"values_source_type": null}' \
   field_ids:='[58, 3]' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        [
            "Doohickey"
        ],
        [
            "Gadget"
        ],
        [
            "Gizmo"
        ],
        [
            "Widget"
        ],
        [
            null
        ],
        [
            "Facebook"
        ],
        [
            "Google"
        ],
        [
            "Invite"
        ],
        [
            "Twitter"
        ]
    ]
}
```

#### FIELD SEARCH

```
❯ http post "localhost:3000/api/dataset/parameter/search/g" \
   parameter:='{"values_source_type": null}' \
   field_ids:='[58, 3]' \
   Cookie:$SESSION -pb
{
    "has_more_values": true,
    "values": [
        [
            "Gadget"
        ],
        [
            "Gizmo"
        ],
        [
            "Widget"
        ],
        [
            "Google"
        ]
    ]
}
```

dpsutton committed:
Get values from unsaved parameters

When hooking up parameters for unsaved dashboards or questions, we need
to get field values.

New endpoint on `api/dataset/parameter/values` and `search`

STATIC LIST VALUES

```
❯ http post "localhost:3000/api/dataset/parameter/values" \
   parameter:='{"values_source_type": "static-list", "values_source_config": {"values": ["foo", "bar"]}}' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        "foo",
        "bar"
    ]
}
```

STATIC LIST SEARCH

```
❯ http post "localhost:3000/api/dataset/parameter/search?query=fo" \
   parameter:='{"values_source_type": "static-list", "values_source_config": {"values": ["foo", "bar"]}}' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        "foo"
    ]
}
```

CARD VALUES

```
❯ http post "localhost:3000/api/dataset/parameter/values" \
   parameter:='{"values_source_type": "card", "values_source_config": {"card_id": 4, "value_field": ["field",58,null]}}' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        "Gizmo",
        "Doohickey",
        "Doohickey",
        ...
        "Gizmo",
        "Widget",
        "Gadget"
    ]
}
```

CARD SEARCH

```
❯ http post "localhost:3000/api/dataset/parameter/search?query=ga" \
   parameter:='{"values_source_type": "card", "values_source_config": {"card_id": 4, "value_field": ["field",58,null]}}' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        "Gadget",
        "Gadget",
        "Gadget",
        "Gadget",
        "Gadget",
        ...
        "Gadget",
        "Gadget",
        "Gadget"
    ]
}
```

FIELD VALUES

```
❯ http post "localhost:3000/api/dataset/parameter/values" \
   parameter:='{"values_source_type": null}' \
   field_ids:='[58, 3]' \
   Cookie:$SESSION -pb
{
    "has_more_values": false,
    "values": [
        [
            "Doohickey"
        ],
        [
            "Gadget"
        ],
        [
            "Gizmo"
        ],
        [
            "Widget"
        ],
        [
            null
        ],
        [
            "Facebook"
        ],
        [
            "Google"
        ],
        [
            "Invite"
        ],
        [
            "Twitter"
        ]
    ]
}
```

```
❯ http post "localhost:3000/api/dataset/parameter/search?query=g" \
   parameter:='{"values_source_type": null}' \
   field_ids:='[58, 3]' \
   Cookie:$SESSION -pb
{
    "has_more_values": true,
    "values": [
        [
            "Gadget"
        ],
        [
            "Gizmo"
        ],
        [
            "Widget"
        ],
        [
            "Google"
        ]
    ]
}
```

dpsutton committed:
tests

dpsutton committed:
extract some shared functionality, docstrings

dpsutton committed:
make schemas private since they don't really need docstrings

dpsutton committed:
suppress clj-kondo deprecated warnings

dpsutton committed:
undo circular requires api.card <--> api.dataset

need field information. moved into api.field for now. But i think there
are 4 or 5 different ways to get field values. Really need to create an
issue to untangle these and crack the knot.

calherries committed:
Change now test to use day, not month for datetime arithmetic (#27943)

ranquild commented on src/metabase/api/dataset.clj:
```diff
@@ -172,4 +175,56 @@
+    (parameter-values parameter field_ids nil-query)))
+
+#_{:clj-kondo/ignore [:deprecated-var]}
+(api/defendpoint-schema POST "/parameter/search"
```
@dpsutton Could we put the query into the url as everywhere else? e.g. https://github.com/metabase/metabase/blob/d369e9408d0cd45aa372d23e8c67611e893bbaf7/src/metabase/api/public.clj#L484

dpsutton replied:
good catch. Fixed.

dpsutton committed:
Move dataset parameter search into route param, not query param

OLD: /api/dataset/parameter/search?query=foo
NEW: /api/dataset/parameter/search/foo
